algs/thermal-noise-removal/src/thermal_noise_utils.cc

```diff
@@ -248,6 +270,77 @@ device::Matrix<double> BuildNoiseLutForTOPSSLC(Rectangle tile, const ThermalNois
+        const auto start_azim_time = first_line_time + nav.first_azimuth_line * line_time_interval;
+        const auto end_azim_time = first_line_time + nav.last_azimuth_line * line_time_interval;
+        const auto noise_vector_indices =
+            DetermineNoiseVectorIndices(start_azim_time, end_azim_time, thermal_noise_info.noise_range_vectors);
+        if (noise_vector_indices.size() == 0) {
+            throw std::runtime_error("There were no matching noise values from noise vectors for azimuth range [" +
+                                     std::to_string(start_azim_time) + ", " + std::to_string(end_azim_time) +
+                                     "]. Tile properties x:" + std::to_string(tile.x) + " y:" + std::to_string(tile.y) +
+                                     " w:" + std::to_string(tile.width) + " h:" + std::to_string(tile.height));
+        }
+
+        const auto interpolated_range_value_count = nx_max - nx0 + 1;
+        std::vector<std::vector<double>> interpolated_range_vectors(noise_vector_indices.size());
+        std::vector<int> noise_range_vector_line(noise_vector_indices.size());
+        for (int i{0}; i < static_cast<int>(noise_vector_indices.size()); i++) {
+            const auto& noise_range_vector = thermal_noise_info.noise_range_vectors.at(noise_vector_indices.at(i));
+            noise_range_vector_line[i] = noise_range_vector.line;
+
+            interpolated_range_vectors.at(i) = std::vector<double>(interpolated_range_value_count);
+            FillRangeNoiseWithInterpolatedValues(noise_range_vector, nx0, nx_max, interpolated_range_vectors.at(i));
+        }
+
+        std::vector<double> interpolated_azimuth_vector(ny_max - ny0 + 1);
+        FillAzimuthNoiseVectorWithInterpolatedValues(nav, ny0, ny_max, interpolated_azimuth_vector);
+        //        std::cout << "Tile size " << tile.width << "x" << tile.height << " for range and azimuth interpolated
+        //        sizes "
+        //                  << interpolated_range_value_count << "x" << interpolated_azimuth_vector.size()
+        //                  << " with interpolated range vector having elements - " << noise_vector_indices.size() <<
+        //                  std::endl;
+        ComputeNoiseMatrix(tile.x, tile.y, nx0, nx_max, ny0, ny_max, noise_range_vector_line,
+                           interpolated_range_vectors, interpolated_azimuth_vector, noise_matrix);
+    }
+
+    if (!has_data) {
+        throw std::runtime_error(
+            "Could not locate a tile for a given range and azimuth set of a product for thermal noise removal "
+            "calculation. Tile properties x:" +
+            std::to_string(tile.x) + " y:" + std::to_string(tile.y) + " w:" + std::to_string(tile.width) +
+            " h:" + std::to_string(tile.height));
+    }
+
+    const auto noise_matrix_dev = device::CreateKernelMatrix<double>(tile.width, tile.height, noise_matrix);
```

Comment:
This breaks streams, because you allocate via cudaMalloc, aka lock all streams for malloc and free. Might be irrelevent for overall performance though.